kuma/wiki/templatetags/ssr.py

```diff
@@ -0,0 +1,128 @@
+    """
+    if ssr:
+        return server_side_render(data)
+    else:
+        return client_side_render(data)
+
+
+def client_side_render(data):
+    """
+    Output an empty <div> and a script with complete state so that
+    the UI can be rendered on the client-side.
+    """
+    # Convert to JSON and be sure the string "</script>" does not appear
+    serializedState = json.dumps(data).replace('</', '<\\/')
+    return OUTPUT_TEMPLATE.format('', serializedState)
+
+
+def _get_renderer():
+    global renderer
+
+    # If we haven't started a renderer process, or if that process
+    # exited for some reason, then start one.
+    if not renderer or renderer.poll() is not None:
+        print "Starting renderer subprocess"
+        renderer = subprocess.Popen(['/usr/bin/env',
+                                     'node',
+                                     '/app/kuma/javascript/dist/ssr.js'],
+                                    stdin=subprocess.PIPE,
+                                    stdout=subprocess.PIPE,
+                                    stderr=sys.stderr)
+    return renderer
+
+
+def server_side_render(data):
+    """
+    Pre-render the React UI to HTML and output it in a <div>, and then
+    also pass the necessary serialized state in a <script> so that
+    React on the client side can sync itself with the pre-rendred HTML.
+
+    If any exceptions are thrown during the server-side rendering, we
+    fall back to client-side rendering instead.
+    """
+    start = time.time()
+
+    # Try server side rendering
+    try:
+        renderer = _get_renderer()
+
+        # To server-side render, we send the JSON-encoded data to
+        # the child process on its stdin stream, and then send
+        # a line of four hyphens to mark the end of the data.
+        renderer.stdin.write(json.dumps(data))
+        renderer.stdin.write('\n----\n')
+        renderer.stdin.flush()
+
+        # The renderer process responds by writing HTML to its stdout
+        # stream. But first, it writes a line of text that it uses
+        # as a boundary (in the MIME multipart sense). So we read the
+        # boundary line, then read lines from the child process until
+        # we see the boundary line again.
+        boundary = renderer.stdout.readline().decode('utf-8')
+        renderedHTML = ''
+        while True:
+            line = renderer.stdout.readline().decode('utf-8')
+            if line == boundary:
+                break
+            renderedHTML += line
+
+        # Even though we've got fully rendered HTML now, we still need
+        # to send the document data along with it so that React can sync
+        # its state on the client side with what is in the HTML.
+        # Fortunately, however, it turns out not to be necessary to
+        # duplicate the biggest parts (the HTML strings) of that data,
+        # so we can delete those from the data now
+        partial = data.copy()
```

Comment:
The, great, code comment above is written as if it's for this `partial = data.copy()` line. I'd rather see something like this:
```python
# Even though we've got fully rendered HTML now, we still need
# to send the document data along with it so that React can sync
# its state on the client side with what is in the HTML.
# Fortunately, however, it turns out not to be necessary to
# duplicate the biggest parts (the HTML strings) of that data,
# so we can delete those from the data now

# We need to make a shallow clone of the data because we ...
partial = data.copy()
```

Also, *why* do you make a shallow clone/copy of it? I don't see it being used by anything else later.

Reply:
The data object is also used in the exception handling code that falls back to client-side rendering, so I'm making a copy in case there is an exception below this point. (Which admittedly is not very likely).

Reply:
I'll update the comment

Reply:
Actually, a better reason for making the copy is that the data dictionary belongs to the caller of this template tag. If we change it here, that change will be seen by anything else in the template that uses it.

Reply:
You're not wrong but it's a cost. A memory cost. And it costs a little in CPU too because of the actual copying of the bytes. 
I think I'm just trying to say; before you commit to this, have you asked yourself if it's worth the cost?

Reply:
It is a trivial cost because this is a shallow copy of one dict. Not making a copy here would be software engineering malpractice because it would be a stealth change to a data structure that might be used elsewhere in the template that calls this template tag.

Reply:
My job is done here I think. I just wanted to force a pause-and-think. 

When I worked on Socorro we fought this hard. A Breakpad crash is read into a Python dict and then it's passed through a long list of "processor classes" that would all get a chance to evaluate and potentially mutate the dict. Or trigger side-effects. We noticed some nasty bugs because of those processor classes needed to *delete* things from the dict, right before storing in Elasticsearch, otherwise it would be too fat on Elasticsearch's disk. So we did `copy.deepcopy(crashobj)` (or whatever it is called). Everything worked locally and in unit tests and in Stage. But then hell broke loose in Prod one day when this change deployed. It blew up because these dicts were big and the deep clone way too expensive. 

We had to go back to the drawing board quick and rewrite it into a more complicated hybrid solution. 

Yeah, deciding between mutation and "immutation" is never a slam dunk.